### ISSUE TITLE
Quantized model much slower than float32

### ISSUE THREAD

**Issue by Daniel-Kelvich** [pytorch/pytorch#42113]
I've tried to quantize a simple model with conv+bn+relu combination but it performs much slower in int8.

## Code To Reproduce

```
import os
import time

import torch.nn as nn
from torch.quantization import QuantStub, DeQuantStub

backend = 'qnnpack'
# backend = 'fbgemm'
import torch
torch.backends.quantized.engine = backend


class DownBlockQ(nn.Module):
    def __init__(self, in_ch, out_ch):
        super().__init__()
        self.quant_input = QuantStub()
        self.dequant_output = DeQuantStub()

        self.conv1 = nn.Conv2d(in_ch, in_ch, 4, stride=2, padding=1, groups=in_ch)
        self.bn1 = nn.BatchNorm2d(in_ch)
        self.relu1 = nn.ReLU()

        self.conv2 = nn.Conv2d(in_ch, out_ch, 1)
        self.bn2 = nn.BatchNorm2d(out_ch)
        self.relu2 = nn.ReLU()

    def forward(self, x):
        # x = self.quant_input(x)
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu1(x)

        x = self.conv2(x)
        x = self.bn2(x)
        x = self.relu2(x)
        # x = self.dequant_output(x)
        return x

    def fuse_model(self):
        torch.quantization.fuse_modules(self, ['conv1', 'bn1', 'relu1'], inplace=True)
        torch.quantization.fuse_modules(self, ['conv2', 'bn2', 'relu2'], inplace=True)


class Model(nn.Module):
    def __init__(self, filters=22):
        super().__init__()
        self.quant_input = QuantStub()
        self.dequant_output = DeQuantStub()

        self.db1 = DownBlockQ(filters * 1, filters * 2)  # 128
        self.db2 = DownBlockQ(filters * 2, filters * 4)  # 64
        self.db3 = DownBlockQ(filters * 4, filters * 8)  # 32

    def forward(self, x):
        x = self.quant_input(x)
        x = self.db1(x)
        x = self.db2(x)
        x = self.db3(x)
        x = self.dequant_output(x)
        return x


def fuse_model(model):
    if hasattr(model, 'fuse_model'):
        model.fuse_model()

    for p in list(model.modules())[1:]:
        fuse_model(p)


def print_size_of_model(model):
    torch.save(model.state_dict(), "temp.p")
    print('Size (MB):', os.path.getsize("temp.p") / 1e6)
    os.remove('temp.p')


def benchmark(func, iters=10, *args):
    t1 = time.time()
    for _ in range(iters):
        res = func(*args)
    print(f'{((time.time() - t1) / iters):.6f} sec')
    return res


def quantize():
    dummy = torch.rand(1, 22, 256, 256)
    # model = DownBlockQ(22 * 1, 22 * 2)
    model = Model(filters=22)
    model = model.eval()
    print("Before quantization")
    print_size_of_model(model)

    benchmark(model, 20, dummy)
    # print(model)
    fuse_model(model)

    model.qconfig = torch.quantization.get_default_qconfig(backend)
    # print(model.qconfig)
    torch.quantization.prepare(model, inplace=True)
    torch.quantization.convert(model, inplace=True)

    # print(model)
    print("After quantization")
    print_size_of_model(model)
    benchmark(model, 20, dummy)
    # torch.jit.script(model).save('models/model_scripted.pt')


if __name__ == '__main__':
    quantize()
```

## Expected behavior

Int8 model to be 2-3 times faster than float32. 

## Environment

PyTorch version: 1.7.0.dev20200727
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 20.04 LTS
GCC version: (Ubuntu 8.4.0-3ubuntu2) 8.4.0
CMake version: version 3.16.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 440.100
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.19.0
[pip3] torch==1.7.0.dev20200727
[pip3] torchvision==0.8.0.dev20200727
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

**Comment by ms-lilibeth** [comment]
I'm facing the same issue
Profiling results before quantization:
```
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                        Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
mkldnn_convolution          50.14%           289.856ms        50.14%           289.856ms        252.049us        1150             
native_batch_norm           21.44%           123.962ms        21.44%           123.962ms        179.654us        690              
sigmoid                     8.76%            50.662ms         8.76%            50.662ms         54.476us         930              
mul                         6.04%            34.923ms         6.04%            34.923ms         37.959us         920              
constant_pad_nd             5.26%            30.431ms         5.26%            30.431ms         126.798us        240    
```

After:
```
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                         Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
quantized::conv2d            75.36%           1.443s           79.00%           1.512s           1.315ms          1150             
quantized::batch_norm        5.25%            100.482ms        5.83%            111.571ms        161.697us        690              
dequantize                   4.33%            82.916ms         4.33%            82.916ms         71.479us         1160             
sigmoid                      2.97%            56.954ms         2.97%            56.954ms         61.241us         930              
contiguous                   2.82%            54.051ms         3.20%            61.274ms         26.411us         2320             
```

I seems that float32 convolution optimized for Intel CPU is working faster than int8 convolution.

**Comment by vkuzo** [comment]
@dskhudia , thoughts on if this is a KP?

**Comment by vkuzo** [comment]
@Daniel-Kelvich , would you mind sharing what kind of CPU you are running this on?

**Comment by dskhudia** [comment]
Discussion is here: https://discuss.pytorch.org/t/slow-quantization/90954

**Comment by Daniel-Kelvich** [comment]
@vkuzo It's  `Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz`

**Comment by kaede10263** [comment]
I execute the code of the official website. But I got this result.My CPU is Intel(R) Core(TM) i9-10900 CPU @ 2.80GH.

Elapsed time:  10 ms
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                        model_inference         5.67%     238.151ms       100.00%        4.201s        4.201s             1  
                                                forward         0.15%       6.284ms        53.56%        2.250s     450.045ms             5  
                                           aten::conv2d         0.06%       2.451ms        42.23%        1.774s       6.824ms           260  
                                      aten::convolution         0.05%       1.919ms        42.17%        1.772s       6.815ms           260  
                                     aten::_convolution         0.08%       3.454ms        42.13%        1.770s       6.808ms           260  
                               aten::mkldnn_convolution        41.94%        1.762s        42.05%        1.767s       6.794ms           260  
enumerate(DataLoader)#_SingleProcessDataLoaderIter._...        38.83%        1.632s        40.76%        1.713s     285.440ms             6  
                                        aten::clamp_min        10.48%     440.367ms        20.93%     879.171ms       2.512ms           350  
                                             aten::relu         0.06%       2.375ms        10.55%     443.149ms       2.532ms           175  
                                              aten::add         0.54%      22.780ms         0.54%      22.780ms     455.600us            50  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 4.201s

Elapsed time:  40 ms
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                        model_inference        83.32%        9.723s       100.00%       11.670s       11.670s             1  
enumerate(DataLoader)#_SingleProcessDataLoaderIter._...        13.72%        1.601s        14.39%        1.679s     279.834ms             6  
                                                forward         0.02%       2.634ms         2.29%     267.627ms      53.525ms             5  
                                 quantized::conv2d_relu         1.61%     188.128ms         1.66%     193.499ms       1.106ms           175  
                                      quantized::conv2d         0.42%      49.536ms         0.43%      50.068ms     589.035us            85  
                                            aten::copy_         0.21%      24.924ms         0.21%      24.928ms      16.509us          1510  
                                            aten::stack         0.00%     324.000us         0.16%      18.564ms       3.094ms             6  
                                              aten::cat         0.00%      45.000us         0.15%      17.385ms       2.897ms             6  
                                             aten::_cat         0.15%      17.274ms         0.15%      17.340ms       2.890ms             6  
                                       aten::contiguous         0.02%       2.348ms         0.13%      15.643ms      51.289us           305  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 11.670s

code : 
import torch
import torchvision.models as models
from torch.profiler import profile, record_function, ProfilerActivity
import torch.profiler
import numpy as np
import torch
import torch.nn as nn
import torchvision
from torch.utils.data import DataLoader
from torchvision import datasets
import torchvision.transforms as transforms
import os
import time
import sys
import torch.quantization

data_path = '~/.data/imagenet'
saved_model_dir = 'data/'
# float_model_file = 'mobilenet_pretrained_float.pth'
scripted_float_model_file = 'mobilenet_quantization_scripted.pth'
scripted_quantized_model_file = 'mobilenet_quantization_scripted_quantized.pth'
train_batch_size = 30
eval_batch_size = 50


def run_benchmark(model_file, img_loader):
    elapsed = 0
    model = torch.jit.load(model_file).to('cpu')
    model.eval()
    num_batches = 5
    with torch.profiler.profile(activities = [ProfilerActivity.CPU], record_shapes = True) as prof:
        with record_function("model_inference"):
            for i, (images, target) in enumerate(img_loader):
                if i < num_batches:
                    start = time.time()
                    output = model(images)
                    end = time.time()
                    elapsed = elapsed + (end-start)

                else:
                    break
        num_images = images.size()[0] * num_batches

        print('Elapsed time: %3.0f ms' % (elapsed/num_images*1000))
    print(prof.key_averages().table(sort_by = "cpu_time_total", row_limit = 10))
    return elapsed

def prepare_data_loaders(data_path):

    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     std=[0.229, 0.224, 0.225])
    dataset = torchvision.datasets.ImageNet(
           data_path, split="train",
        transform=transforms.Compose([
                   transforms.RandomResizedCrop(224),
                   transforms.RandomHorizontalFlip(),
                   transforms.ToTensor(),
                   normalize,
               ]))
    dataset_test = torchvision.datasets.ImageNet(
          data_path, split="val",
              transform=transforms.Compose([
                  transforms.Resize(256),
                  transforms.CenterCrop(224),
                  transforms.ToTensor(),
                  normalize,
              ]))

    train_sampler = torch.utils.data.RandomSampler(dataset)
    test_sampler = torch.utils.data.SequentialSampler(dataset_test)

    data_loader = torch.utils.data.DataLoader(
        dataset, batch_size=train_batch_size, # pin_memory ="True",
        sampler=train_sampler)

    data_loader_test = torch.utils.data.DataLoader(
        dataset_test, batch_size=eval_batch_size, # pin_memory ="True",
        sampler=test_sampler)
    
    return data_loader, data_loader_test


data_loader, data_loader_test = prepare_data_loaders(data_path)
run_benchmark(saved_model_dir + scripted_float_model_file, data_loader_test)
run_benchmark(saved_model_dir + scripted_quantized_model_file, data_loader_test)

thanks~

**Comment by jerryzh168** [comment]
is this still relevant? we have many changes in quantization since the time of the issue, I'll close for now and please move to the new flow and try again: https://pytorch.org/docs/main/quantization.html#prototype-pytorch-2-export-quantization (you can use https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html for server CPU)
